File: swift/Sources/ForySwift/AnySerializer.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

private enum DynamicAnyMapHeader {
    static let keyNull: UInt8 = 0b0000_0010
    static let declaredKeyType: UInt8 = 0b0000_0100
    static let valueNull: UInt8 = 0b0001_0000
}

public struct ForyAnyNullValue {
    public init() {}
}

private struct DynamicAnyValue: Serializer {
    var value: Any = ForyAnyNullValue()

    init(_ value: Any) {
        self.value = value
    }

    static func foryDefault() -> DynamicAnyValue {
        DynamicAnyValue(ForyAnyNullValue())
    }

    static var staticTypeId: ForyTypeId {
        .unknown
    }

    static var isNullableType: Bool {
        true
    }

    static var isReferenceTrackableType: Bool {
        true
    }

    var foryIsNone: Bool {
        value is ForyAnyNullValue
    }

    static func wrapped(_ value: Any?) -> DynamicAnyValue {
        guard let value else {
            return .foryDefault()
        }
        guard let unwrapped = unwrapOptionalAny(value) else {
            return .foryDefault()
        }
        if unwrapped is NSNull {
            return .foryDefault()
        }
        return DynamicAnyValue(unwrapped)
    }

    func anyValue() -> Any? {
        foryIsNone ? nil : value
    }

    func anyValueForCollection() -> Any {
        foryIsNone ? NSNull() : value
    }

    func foryWriteData(_ context: WriteContext, hasGenerics: Bool) throws {
        if foryIsNone {
            return
        }
        try writeAnyPayload(value, context: context, hasGenerics: hasGenerics)
    }

    static func foryReadData(_ context: ReadContext) throws -> DynamicAnyValue {
        guard let typeInfo = context.dynamicTypeInfo(for: Self.self) else {
            throw ForyError.invalidData("dynamic Any value requires type info")
        }
        context.clearDynamicTypeInfo(for: Self.self)
        if typeInfo.wireTypeID == .none {
            return .foryDefault()
        }
        return DynamicAnyValue(try context.typeResolver.readDynamicValue(typeInfo: typeInfo, context: context))
    }

    static func foryWriteTypeInfo(_ context: WriteContext) throws {
        _ = context
        throw ForyError.invalidData("dynamic Any value type info is runtime-only")
    }

    func foryWriteTypeInfo(_ context: WriteContext) throws {
        if foryIsNone {
            context.writer.writeUInt8(UInt8(truncatingIfNeeded: ForyTypeId.none.rawValue))
            return
        }
        try writeAnyTypeInfo(value, context: context)
    }

    static func foryReadTypeInfo(_ context: ReadContext) throws {
        let typeInfo = try context.typeResolver.readDynamicTypeInfo(context: context)
        context.setDynamicTypeInfo(for: Self.self, typeInfo)
    }

    func foryWrite(
        _ context: WriteContext,
        refMode: RefMode,
        writeTypeInfo: Bool,
        hasGenerics: Bool
    ) throws {
        if refMode != .none {
            if foryIsNone {
                context.writer.writeInt8(RefFlag.null.rawValue)
                return
            }
            if refMode == .tracking, anyValueIsReferenceTrackable(value), let object = value as AnyObject? {
                if context.refWriter.tryWriteReference(writer: context.writer, object: object) {
                    return
                }
            } else {
                context.writer.writeInt8(RefFlag.notNullValue.rawValue)
            }
        }

        if writeTypeInfo {
            try foryWriteTypeInfo(context)
        }
        try foryWriteData(context, hasGenerics: hasGenerics)
    }

    static func foryRead(
        _ context: ReadContext,
        refMode: RefMode,
        readTypeInfo: Bool
    ) throws -> DynamicAnyValue {
        if refMode != .none {
            let rawFlag = try context.reader.readInt8()
            guard let flag = RefFlag(rawValue: rawFlag) else {
                throw ForyError.refError("invalid ref flag \(rawFlag)")
            }

            switch flag {
            case .null:
                return .foryDefault()
            case .ref:
                let refID = try context.reader.readVarUInt32()
                let referenced = try context.refReader.readRefValue(refID)
                guard let value = referenced as? DynamicAnyValue else {
                    throw ForyError.refError("ref_id \(refID) has unexpected runtime type")
                }
                return value
            case .refValue:
                let reservedRefID = context.refReader.reserveRefID()
                context.pushPendingReference(reservedRefID)
                if readTypeInfo {
                    try foryReadTypeInfo(context)
                }
                let value = try foryReadData(context)
                context.finishPendingReferenceIfNeeded(value)
                context.popPendingReference()
                return value
            case .notNullValue:
                break
            }
        }

        if readTypeInfo {
            try foryReadTypeInfo(context)
        }
        return try foryReadData(context)
    }
}

private func unwrapOptionalAny(_ value: Any) -> Any? {
    let mirror = Mirror(reflecting: value)
    guard mirror.displayStyle == .optional else {
        return value
    }
    guard let (_, child) = mirror.children.first else {
        return nil
    }
    return child
}

private func anyValueIsReferenceTrackable(_ value: Any) -> Bool {
    guard let serializer = value as? any Serializer else {
        return false
    }
    return type(of: serializer).isReferenceTrackableType
}

private func writeAnyTypeInfo(_ value: Any, context: WriteContext) throws {
    if let serializer = value as? any Serializer {
        try serializer.foryWriteTypeInfo(context)
        return
    }

    if value is [Any] {
        context.writer.writeUInt8(UInt8(truncatingIfNeeded: ForyTypeId.list.rawValue))
        return
    }
    if value is [String: Any] || value is [Int32: Any] {
        context.writer.writeUInt8(UInt8(truncatingIfNeeded: ForyTypeId.map.rawValue))
        return
    }

    throw ForyError.invalidData("unsupported dynamic Any runtime type \(type(of: value))")
}

private func writeAnyPayload(_ value: Any, context: WriteContext, hasGenerics: Bool) throws {
    if let serializer = value as? any Serializer {
        try serializer.foryWriteData(context, hasGenerics: hasGenerics)
        return
    }
    if let list = value as? [Any] {
        try writeAnyList(list, context: context, refMode: .none, hasGenerics: hasGenerics)
        return
    }
    if let map = value as? [String: Any] {
        // Always include key type info for dynamic map payload.
        try writeStringAnyMap(map, context: context, refMode: .none, hasGenerics: false)
        return
    }
    if let map = value as? [Int32: Any] {
        // Always include key type info for dynamic map payload.
        try writeInt32AnyMap(map, context: context, refMode: .none, hasGenerics: false)
        return
    }
    throw ForyError.invalidData("unsupported dynamic Any runtime type \(type(of: value))")
}

public func castAnyDynamicValue<T>(_ value: Any?, to type: T.Type) throws -> T {
    guard let typed = value as? T else {
        throw ForyError.invalidData("cannot cast dynamic Any value to \(type)")
    }
    return typed
}

public func writeAny(
    _ value: Any?,
    context: WriteContext,
    refMode: RefMode,
    writeTypeInfo: Bool = true,
    hasGenerics: Bool = false
) throws {
    try DynamicAnyValue.wrapped(value).foryWrite(
        context,
        refMode: refMode,
        writeTypeInfo: writeTypeInfo,
        hasGenerics: hasGenerics
    )
}

public func readAny(
    context: ReadContext,
    refMode: RefMode,
    readTypeInfo: Bool = true
) throws -> Any? {
    try DynamicAnyValue.foryRead(context, refMode: refMode, readTypeInfo: readTypeInfo).anyValue()
}

public func writeAnyList(
    _ value: [Any]?,
    context: WriteContext,
    refMode: RefMode,
    writeTypeInfo: Bool = false,
    hasGenerics: Bool = true
) throws {
    let wrapped = value?.map { DynamicAnyValue.wrapped($0) }
    try wrapped.foryWrite(
        context,
        refMode: refMode,
        writeTypeInfo: writeTypeInfo,
        hasGenerics: hasGenerics
    )
}

public func readAnyList(
    context: ReadContext,
    refMode: RefMode,
    readTypeInfo: Bool = false
) throws -> [Any]? {
    let wrapped: [DynamicAnyValue]? = try [DynamicAnyValue]?.foryRead(
        context,
        refMode: refMode,
        readTypeInfo: readTypeInfo
    )
    return wrapped?.map { $0.anyValueForCollection() }
}

public func writeStringAnyMap(
    _ value: [String: Any]?,
    context: WriteContext,
    refMode: RefMode,
    writeTypeInfo: Bool = false,
    hasGenerics: Bool = true
) throws {
    let wrapped = value?.reduce(into: [String: DynamicAnyValue]()) { result, pair in
        result[pair.key] = DynamicAnyValue.wrapped(pair.value)
    }
    try wrapped.foryWrite(
        context,
        refMode: refMode,
        writeTypeInfo: writeTypeInfo,
        hasGenerics: hasGenerics
    )
}

public func readStringAnyMap(
    context: ReadContext,
    refMode: RefMode,
    readTypeInfo: Bool = false
) throws -> [String: Any]? {
    let wrapped: [String: DynamicAnyValue]? = try [String: DynamicAnyValue]?.foryRead(
        context,
        refMode: refMode,
        readTypeInfo: readTypeInfo
    )
    guard let wrapped else {
        return nil
    }
    var map: [String: Any] = [:]
    map.reserveCapacity(wrapped.count)
    for pair in wrapped {
        map[pair.key] = pair.value.anyValueForCollection()
    }
    return map
}

public func writeInt32AnyMap(
    _ value: [Int32: Any]?,
    context: WriteContext,
    refMode: RefMode,
    writeTypeInfo: Bool = false,
    hasGenerics: Bool = true
) throws {
    let wrapped = value?.reduce(into: [Int32: DynamicAnyValue]()) { result, pair in
        result[pair.key] = DynamicAnyValue.wrapped(pair.value)
    }
    try wrapped.foryWrite(
        context,
        refMode: refMode,
        writeTypeInfo: writeTypeInfo,
        hasGenerics: hasGenerics
    )
}

public func readInt32AnyMap(
    context: ReadContext,
    refMode: RefMode,
    readTypeInfo: Bool = false
) throws -> [Int32: Any]? {
    let wrapped: [Int32: DynamicAnyValue]? = try [Int32: DynamicAnyValue]?.foryRead(
        context,
        refMode: refMode,
        readTypeInfo: readTypeInfo
    )
    guard let wrapped else {
        return nil
    }
    var map: [Int32: Any] = [:]
    map.reserveCapacity(wrapped.count)
    for pair in wrapped {
        map[pair.key] = pair.value.anyValueForCollection()
    }
    return map
}

func readDynamicAnyMapValue(context: ReadContext) throws -> Any {
    let mapStart = context.reader.getCursor()
    let keyTypeID = try peekDynamicMapKeyTypeID(context: context)
    context.reader.setCursor(mapStart)

    switch keyTypeID {
    case .int32, .varint32:
        return try readInt32AnyMap(context: context, refMode: .none) ?? [:]
    case nil, .string:
        return try readStringAnyMap(context: context, refMode: .none) ?? [:]
    default:
        throw ForyError.invalidData("unsupported dynamic map key type \(String(describing: keyTypeID))")
    }
}

private func peekDynamicMapKeyTypeID(context: ReadContext) throws -> ForyTypeId? {
    let start = context.reader.getCursor()
    defer {
        context.reader.setCursor(start)
    }

    let length = Int(try context.reader.readVarUInt32())
    if length == 0 {
        return nil
    }

    let header = try context.reader.readUInt8()
    let keyNull = (header & DynamicAnyMapHeader.keyNull) != 0
    let keyDeclared = (header & DynamicAnyMapHeader.declaredKeyType) != 0
    let valueNull = (header & DynamicAnyMapHeader.valueNull) != 0

    if keyDeclared {
        return nil
    }
    if keyNull {
        return nil
    }

    if !valueNull {
        _ = try context.reader.readUInt8()
    }

    let rawTypeID = try context.reader.readVarUInt32()
    return ForyTypeId(rawValue: rawTypeID)
}

File: csharp/src/Fory/AnySerializer.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

namespace Apache.Fory;

public sealed class DynamicAnyObjectSerializer : Serializer<object?>
{
    public override TypeId StaticTypeId => TypeId.Unknown;
    public override bool IsNullableType => true;
    public override bool IsReferenceTrackableType => true;
    public override object? DefaultValue => null;
    public override bool IsNone(in object? value) => value is null;

    public override void WriteData(WriteContext context, in object? value, bool hasGenerics)
    {
        if (IsNone(value))
        {
            return;
        }

        DynamicAnyCodec.WriteAnyPayload(value!, context, hasGenerics);
    }

    public override object? ReadData(ReadContext context)
    {
        DynamicTypeInfo? dynamicTypeInfo = context.DynamicTypeInfo(typeof(object));
        if (dynamicTypeInfo is null)
        {
            throw new InvalidDataException("dynamic Any value requires type info");
        }

        return context.TypeResolver.ReadDynamicValue(dynamicTypeInfo, context);
    }

    public override void WriteTypeInfo(WriteContext context)
    {
        throw new InvalidDataException("dynamic Any value type info is runtime-only");
    }

    public override void ReadTypeInfo(ReadContext context)
    {
        DynamicTypeInfo typeInfo = context.TypeResolver.ReadDynamicTypeInfo(context);
        context.SetDynamicTypeInfo(typeof(object), typeInfo);
    }

    public override void Write(WriteContext context, in object? value, RefMode refMode, bool writeTypeInfo, bool hasGenerics)
    {
        if (refMode != RefMode.None)
        {
            if (IsNone(value))
            {
                context.Writer.WriteInt8((sbyte)RefFlag.Null);
                return;
            }

            bool wroteTrackingRefFlag = false;
            if (refMode == RefMode.Tracking && AnyValueIsReferenceTrackable(value!, context.TypeResolver))
            {
                if (context.RefWriter.TryWriteReference(context.Writer, value!))
                {
                    return;
                }

                wroteTrackingRefFlag = true;
            }

            if (!wroteTrackingRefFlag)
            {
                context.Writer.WriteInt8((sbyte)RefFlag.NotNullValue);
            }
        }

        if (writeTypeInfo)
        {
            DynamicAnyCodec.WriteAnyTypeInfo(value!, context);
        }

        WriteData(context, value, hasGenerics);
    }

    public override object? Read(ReadContext context, RefMode refMode, bool readTypeInfo)
    {
        if (refMode != RefMode.None)
        {
            sbyte rawFlag = context.Reader.ReadInt8();
            RefFlag flag = (RefFlag)rawFlag;
            switch (flag)
            {
                case RefFlag.Null:
                    return null;
                case RefFlag.Ref:
                {
                    uint refId = context.Reader.ReadVarUInt32();
                    return context.RefReader.ReadRefValue(refId);
                }
                case RefFlag.RefValue:
                {
                    uint reservedRefId = context.RefReader.ReserveRefId();
                    context.RefReader.PushPendingReference(reservedRefId);
                    if (readTypeInfo)
                    {
                        ReadTypeInfo(context);
                    }

                    object? value = ReadData(context);
                    if (readTypeInfo)
                    {
                        context.ClearDynamicTypeInfo(typeof(object));
                    }

                    context.RefReader.FinishPendingReferenceIfNeeded(value);
                    context.RefReader.PopPendingReference();
                    return value;
                }
                case RefFlag.NotNullValue:
                    break;
                default:
                    throw new RefException($"invalid ref flag {rawFlag}");
            }
        }

        if (readTypeInfo)
        {
            ReadTypeInfo(context);
        }

        object? result = ReadData(context);
        if (readTypeInfo)
        {
            context.ClearDynamicTypeInfo(typeof(object));
        }

        return result;
    }

    private static bool AnyValueIsReferenceTrackable(object value, TypeResolver typeResolver)
    {
        TypeInfo typeInfo = typeResolver.GetTypeInfo(value.GetType());
        return typeInfo.IsReferenceTrackableType;
    }
}

public static class DynamicAnyCodec
{
    internal static void WriteAnyTypeInfo(object value, WriteContext context)
    {
        if (DynamicContainerCodec.TryGetTypeId(value, out TypeId containerTypeId))
        {
            context.Writer.WriteUInt8((byte)containerTypeId);
            return;
        }

        if (TryWriteKnownTypeInfo(value, context))
        {
            return;
        }

        TypeInfo typeInfo = context.TypeResolver.GetTypeInfo(value.GetType());
        typeInfo.WriteTypeInfo(context);
    }

    public static object? CastAnyDynamicValue(object? value, Type targetType)
    {
        if (value is null)
        {
            if (targetType == typeof(object))
            {
                return null;
            }

            if (!targetType.IsValueType || Nullable.GetUnderlyingType(targetType) is not null)
            {
                return null;
            }

            throw new InvalidDataException($"cannot cast null dynamic Any value to non-nullable {targetType}");
        }

        if (targetType.IsInstanceOfType(value))
        {
            return value;
        }

        throw new InvalidDataException($"cannot cast dynamic Any value to {targetType}");
    }

    public static void WriteAny(WriteContext context, object? value, RefMode refMode, bool writeTypeInfo = true, bool hasGenerics = false)
    {
        context.TypeResolver.GetSerializer<object?>().Write(context, value, refMode, writeTypeInfo, hasGenerics);
    }

    public static object? ReadAny(ReadContext context, RefMode refMode, bool readTypeInfo = true)
    {
        return context.TypeResolver.GetSerializer<object?>().Read(context, refMode, readTypeInfo);
    }

    public static void WriteAnyPayload(object value, WriteContext context, bool hasGenerics)
    {
        if (DynamicContainerCodec.TryWritePayload(value, context, hasGenerics))
        {
            return;
        }

        if (TryWriteKnownPayload(value, context))
        {
            return;
        }

        TypeInfo typeInfo = context.TypeResolver.GetTypeInfo(value.GetType());
        typeInfo.WriteDataObject(context, value, hasGenerics);
    }

    private static bool TryWriteKnownTypeInfo(object value, WriteContext context)
    {
        switch (value)
        {
            case bool:
                context.Writer.WriteUInt8((byte)TypeId.Bool);
                return true;
            case sbyte:
                context.Writer.WriteUInt8((byte)TypeId.Int8);
                return true;
            case short:
                context.Writer.WriteUInt8((byte)TypeId.Int16);
                return true;
            case int:
                context.Writer.WriteUInt8((byte)TypeId.VarInt32);
                return true;
            case long:
                context.Writer.WriteUInt8((byte)TypeId.VarInt64);
                return true;
            case byte:
                context.Writer.WriteUInt8((byte)TypeId.UInt8);
                return true;
            case ushort:
                context.Writer.WriteUInt8((byte)TypeId.UInt16);
                return true;
            case uint:
                context.Writer.WriteUInt8((byte)TypeId.VarUInt32);
                return true;
            case ulong:
                context.Writer.WriteUInt8((byte)TypeId.VarUInt64);
                return true;
            case float:
                context.Writer.WriteUInt8((byte)TypeId.Float32);
                return true;
            case double:
                context.Writer.WriteUInt8((byte)TypeId.Float64);
                return true;
            case string:
                context.Writer.WriteUInt8((byte)TypeId.String);
                return true;
            case byte[]:
                context.Writer.WriteUInt8((byte)TypeId.Binary);
                return true;
            case bool[]:
                context.Writer.WriteUInt8((byte)TypeId.BoolArray);
                return true;
            case sbyte[]:
                context.Writer.WriteUInt8((byte)TypeId.Int8Array);
                return true;
            case short[]:
                context.Writer.WriteUInt8((byte)TypeId.Int16Array);
                return true;
            case int[]:
                context.Writer.WriteUInt8((byte)TypeId.Int32Array);
                return true;
            case long[]:
                context.Writer.WriteUInt8((byte)TypeId.Int64Array);
                return true;
            case ushort[]:
                context.Writer.WriteUInt8((byte)TypeId.UInt16Array);
                return true;
            case uint[]:
                context.Writer.WriteUInt8((byte)TypeId.UInt32Array);
                return true;
            case ulong[]:
                context.Writer.WriteUInt8((byte)TypeId.UInt64Array);
                return true;
            case float[]:
                context.Writer.WriteUInt8((byte)TypeId.Float32Array);
                return true;
            case double[]:
                context.Writer.WriteUInt8((byte)TypeId.Float64Array);
                return true;
            case DateOnly:
                context.Writer.WriteUInt8((byte)TypeId.Date);
                return true;
            case DateTimeOffset:
            case DateTime:
                context.Writer.WriteUInt8((byte)TypeId.Timestamp);
                return true;
            case TimeSpan:
                context.Writer.WriteUInt8((byte)TypeId.Duration);
                return true;
            default:
                return false;
        }
    }

    private static bool TryWriteKnownPayload(object value, WriteContext context)
    {
        switch (value)
        {
            case bool v:
                context.Writer.WriteUInt8(v ? (byte)1 : (byte)0);
                return true;
            case sbyte v:
                context.Writer.WriteInt8(v);
                return true;
            case short v:
                context.Writer.WriteInt16(v);
                return true;
            case int v:
                context.Writer.WriteVarInt32(v);
                return true;
            case long v:
                context.Writer.WriteVarInt64(v);
                return true;
            case byte v:
                context.Writer.WriteUInt8(v);
                return true;
            case ushort v:
                context.Writer.WriteUInt16(v);
                return true;
            case uint v:
                context.Writer.WriteVarUInt32(v);
                return true;
            case ulong v:
                context.Writer.WriteVarUInt64(v);
                return true;
            case float v:
                context.Writer.WriteFloat32(v);
                return true;
            case double v:
                context.Writer.WriteFloat64(v);
                return true;
            case string v:
                StringSerializer.WriteString(context, v);
                return true;
            case DateOnly v:
                TimeCodec.WriteDate(context, v);
                return true;
            case DateTimeOffset v:
                TimeCodec.WriteTimestamp(context, v);
                return true;
            case DateTime v:
                TimeCodec.WriteTimestamp(context, TimeCodec.ToDateTimeOffset(v));
                return true;
            case TimeSpan v:
                TimeCodec.WriteDuration(context, v);
                return true;
            case byte[] v:
                context.Writer.WriteVarUInt32((uint)v.Length);
                context.Writer.WriteBytes(v);
                return true;
            case bool[] v:
                context.Writer.WriteVarUInt32((uint)v.Length);
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteUInt8(v[i] ? (byte)1 : (byte)0);
                }
                return true;
            case sbyte[] v:
                context.Writer.WriteVarUInt32((uint)v.Length);
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteInt8(v[i]);
                }
                return true;
            case short[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 2));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteInt16(v[i]);
                }
                return true;
            case int[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 4));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteInt32(v[i]);
                }
                return true;
            case long[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 8));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteInt64(v[i]);
                }
                return true;
            case ushort[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 2));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteUInt16(v[i]);
                }
                return true;
            case uint[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 4));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteUInt32(v[i]);
                }
                return true;
            case ulong[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 8));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteUInt64(v[i]);
                }
                return true;
            case float[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 4));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteFloat32(v[i]);
                }
                return true;
            case double[] v:
                context.Writer.WriteVarUInt32((uint)(v.Length * 8));
                for (int i = 0; i < v.Length; i++)
                {
                    context.Writer.WriteFloat64(v[i]);
                }
                return true;
            default:
                return false;
        }
    }
}
